call_fd_callback.c

```diff
@@ -0,0 +1,12 @@
+#include <stdint.h>
+
+typedef void (*fd_callback)(void *class_ptr, int);
+
+// bridge to allow a function pointer to be passed to go. That ptr must be called from c.
+void call_fd_callback(void* func_ptr, void* class_ptr, uintptr_t fd)
+{
+  if (func_ptr) {
```

Comment:
High level: what's the case where Go would pass a null for `func_ptr`?

Lower level: can we handle that case a bit more explicitly, either in C or in Go? The way it's currently built, between this being a `void` function and not surfacing anything when `func_ptr` is null, `main.go:136` could be assigning the `error` value in a broader set of cases.

Reply:
hmm thats a good question. I like to check pointers for nil, but in this case it's a "it'll always work or never work" situation. There _should_ never be a case where func_ptr is nil, but if it is C will segfault. 

I don't think its necessary to have this function return a status code, but I think its good to keep the null check in? Maybe in the go code?: 
```
//export Obfs4_get_fd
func Obfs4_get_fd(listener_id int, function unsafe.Pointer, class_ptr unsafe.Pointer) int {
  if function == nil {
    return -1
  }
```

Reply:
Yeah, I think that snippet makes sense here — as long as we actually surface this to the client somehow. Having silent failure here could get _really_ frustrating during debugging.

Reply:
Done. I suspect we will get more feedback when we merge upstream

Reply:
Yup, that's likely, but for now this is really solid work.